rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,654 @@
+
+We need to ensure that:
+
+1. All backend operations are applied in the order they are received, any outdated requests are rejected.
+2. Auth servers know the exact version of a resource requested by clients.
+3. Once a migration has been performed, Auth servers that cannot understand the migrated resource version cannot
+   overwrite the resource.
+4. Migrations are always applied in the correct order and are not skipped.
+5. Migrations can be rolled back without having to manually edit the backend.
+
+#### Resource versioning
+
+The version of a resource MUST be bumped when changes are made to it. Any changes to a resource which can be converted
+into the previous version and do not cause any backward compatibility issues with older Teleport instances only need to
+bump the minor version of a resource. All changes to a resource which would cause backward incompatibility with other
+Auth servers trying to read/write the resource MUST update the major version. All changes to a resource which alter how
+that resource is understood, interpreted, and acted upon by Teleport instances MUST update the major version.
+
+For example, if we originally have a resource like the following at v1:
+
+```proto
+syntax = "proto3";
+
+message Foo {
+  int32 bar = 1;
+}
+```
+
+Then adding a new optional field which was handled appropriately, and defaulted to the correct value by application code
+if empty only requires bumping the version to v1.1 and not require a direct migration.
+
+```diff
+message Foo {
+  int32 bar = 1;
++  string baz = 2;
+}
+```
+
+However, if we were to convert `baz` from a string to `Baz`, that change would not be easily converted into the shape of
+`Foo` at v1.1 and would render the new resource unusable by clients that are only aware of `Foo` v1.1. So, the version
+of the resource must be bumped from v1.1 to v2 to indicate to clients that a breaking change occurred.
+
+```diff
+message Foo {
+  int32 bar = 1;
+- string baz = 2;
++ reserved 2;
++ reserved baz;
++ Baz baz2 = 3;
+}
+
++ message Baz {
++  string qux = 1;
++  int32 quux = 2;
++ }
+```
+
+To date, Auth assumes all client requests are for the version of the resource that Auth is aware of. However, this
+causes problems when a resource version is bumped with a breaking change since it is not guaranteed that all Teleport
+instances in a cluster are running the same version as Auth. The solution has been to downgrade the resource to the
+previous version or alter the resource based on the version of the requester as indicated by the `version` header.
+
+For clients to better communicate which version of a resource they can support there are few options:
+
+1. Add a header that clients must populate with their greatest known version of a resource
+2. Version the API such that the version is implied
+
+For major breaking changes a new version of the API is likely warranted, however for smaller changes to a resource it
```

Comment:
This is starting to feel very complicated - we've basically introduced 3 new versioning mechanisms:

- resource minor version
- resource major version
- API version

Do you think it will become difficult to track things like "role version 8.1 is only supported in API versions 2 and 3, and API version 4 requires role version 9+"